src/shared/specifier.c

```diff
@@ -22,6 +22,7 @@
 #include "specifier.h"
 #include "string-util.h"
 #include "strv.h"
+#include "unit-file.h"
```

Comment:
I'd really like to keep the specifier stuff indepdent of unit files. We use the specifier stuff all over the place after all.

hence, instead of this, why not define a flags field to pass in the `data` parameter, i.e.

```c
typedef enum SpecifierFlags {
        SPECIFIER_ROOT_USER = 1 << 0, /* force user/group resolution to return "root", "0" */
        SPECIFIER_FAIL_USER = 1 << 1, /* make user/group specifiers fail */
} SpecifierFlags;
```

and then add a little helper func that can translate the UnitFileScope to SpecifierFlags, so that you can easily set these flags based on the scope.

(Another idea would be to make these flags an extra parameter all specifier callbacks take, separate from `data`)

Reply:
So… I don't like this approach. There'd be 1:1 mapping of the new enum and UnitFileScope. UnitFileScope is well established and widely used, and adding a new enum would just introduce a layer of abstraction that doesn't simplify anything. And I think it's OK to pass this as `data`. We don't use `data` for anything else, and for those specifiers, we're unlikely to ever do.

The part that is somewhat ugly are the conversions using `PTR_TO_INT` and `INT_TO_PTR`. But afaics, this part we can't get rid of.

Reply:
UnitFileScope is not well established outside of unit file handling. tmpfiles.d, sysusers.d and all the other tools that deal with specifiers cannot make sense of that...

Reply:
I mean, I would be more sympathetic to reusing UnitFileScope if it would be generalized, i.e. InvocationScope or AppScope or so, so that we can reasonably use it in all our tools, after all similar concepts kinda do make sense there.

Reply:
or tldr: I really don#t like the fact that specifier.h now includes unit-file.h.

Reply:
generalizing UnitFileScope as InvocationScope makes a ton of sense btw, we use similar concepts everywhere after all, i.e. when talking about D-Bus busses and stuff

Reply:
New patch at the end now renames it to LookupScope.